Nas1 shot1

PhMueller opened the pull request.

mlindauer commented:
Main problem:
```
module = importlib.import_module("hpolib.benchmarks.surrogates.exploring_openml")
benchmark = getattr(module, "XGBoost_3")(rng=seed)
```
Further issues:
 
- [ ] meta_information is missing (in particular `num_function_evals`, `name`, `f_opt` (can be set to 0))
- [ ] please remove the print statements

mlindauer commented:
Now, it takes really a lot of time. I'm trying to import XGBoost_3 now for roughly half an hour. Unfortunately, there is not output and hence, I don't know what it is doing at all. Could you please some (sparse) output such that a user does not believe that he/she hits an infinite loop?

Is the reproducibility still ensured since the HPO runs every time again? So, can I validate the configuration I found on my local machine again and it will return the same results?

KEggensperger commented on hpolib/benchmarks/surrogates/exploring_openml.py:
```diff
@@ -0,0 +1,515 @@
+        Parameters
+        ----------
+        dataset_id: int
+            Dataset Id as given in Table 2.
```
Table 2? Which Table 2 ...
Could you maybe add a list of valid ids in the meta_data?

KEggensperger replied:
Okay, I just found the list. So this is then rather an Abstract class that should not be used directly. Could you add that somewhere in the comments?

mfeurer replied:
Added this.

KEggensperger commented on hpolib/benchmarks/surrogates/exploring_openml.py:
```diff
@@ -0,0 +1,515 @@
+            surrogate_dir,
+            'surrogate_%s_%d.pkl.gz' % (self.classifier, self.dataset_id),
+        )
+        if os.path.exists(surrogate_file_name):
```
If I run this benchmark once and then use n_splits=10 and I run it again with n_splits=5 will this update the surrogate?

mfeurer replied:
No.

KEggensperger commented on hpolib/benchmarks/surrogates/exploring_openml.py:
```diff
@@ -0,0 +1,515 @@
+    """
+    url = None
+
+    def __init__(self, dataset_id, n_splits=10, rng=None):
```
Could you also add a rebuild flag? (related to my comment below)

mfeurer replied:
Yes, I just added this.

KEggensperger commented on hpolib/benchmarks/surrogates/exploring_openml.py:
```diff
@@ -0,0 +1,515 @@
+            max_features,
+            bootstrap,
+        ])
+        cs.seed(1)
```
This makes constructing the surrogate deterministic, correct?

mfeurer replied:
Correct.

KEggensperger commented on hpolib/benchmarks/surrogates/exploring_openml.py:
```diff
@@ -0,0 +1,515 @@
+            n_iterations += 1
+
+            # Stopping condititions
+            if n_iterations > 40:
```
This should be an input with a default to the benchmark.

mfeurer replied:
I added this, and also use it in the unit tests now.

KEggensperger commented on hpolib/benchmarks/surrogates/exploring_openml.py:
```diff
@@ -0,0 +1,515 @@
+                self.logger.debug('Stopping random search due to good performance!')
+                break
+            elif highest_correlations > 0.999:
+                self.logger.debug('Stopping random search due to good performance!')
```
Where do these numbers come from?

mfeurer replied:
I removed these numbers. These were based on intuition and the results your AAAI paper.

KEggensperger commented on hpolib/benchmarks/surrogates/exploring_openml.py:
```diff
@@ -0,0 +1,515 @@
+        self.impute_with_defaults(x)
+        y = self.regressor.predict(x)
+        y = y[0]
+        return {'function_value': y}
```
Is there no runtime we can return?

mfeurer replied:
I just added the runtime.

KEggensperger commented on hpolib/benchmarks/surrogates/exploring_openml.py:
```diff
@@ -0,0 +1,515 @@
+        return {
+            'num_function_evals': 50,
+            'name': 'Exploring_OpenML',
+            'f_opt': 0,
```
Are you sure about this?

mfeurer replied:
I'd be happy to change this, but given the fact that the class is abstract I can't add a way to access the data and therefore need to assume the lowest possible value. I added two functions to retrieve the empirical min and max which is necessary to compute some metrics.

KEggensperger commented on hpolib/benchmarks/surrogates/exploring_openml.py:
```diff
@@ -0,0 +1,515 @@
+        exec(benchmark_string)
+
+
+if __name__ == '__main__':
```
Why is this here?

mfeurer replied:
For some testing if all methods can be instantiated.

mfeurer commented:
Okay @KEggensperger I think I addressed all your requests.

PhMueller committed:
Add nasbench_1shot1 benchmark

PhMueller committed:
[skip travis] Initial Commit Nas1shot1 Benchmark

PhMueller committed:
[skip travis] Nasbench1shot1: Fix Json Serialization + Encoder

- The ConfigSpace cannot serialize hyperparameter which consists of Tuple(tuple(Any)).
Therefore, I cast them to strings and recast them to tuple later.

- Small improvements in codestyle.

- Also, I have added a JSON Encoder for handling tuple, numpy-types and others.
This is not used atm but will be then integrated in the next big pull request.

PhMueller committed:
[skip travis] Pycodestyle

PhMueller committed:
Add Recipe 1shot1 + (info about incumbents for nasbench_201.py)

PhMueller committed:
[skip travis] Improve changelog.md

KEggensperger commented on extra_requirements/nasbench_1shot1.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "nasbench_1shot1": ["tensorflow==1.15.0","matplotlib","seaborn",
+                      "ConfigSpace","networkx", "tqdm"]
```
the configspace should be already installed?

KEggensperger commented on hpobench/benchmarks/nas/nasbench_201.py:
```diff
@@ -100,6 +100,15 @@ def __init__(self, dataset: str,
 
+        Incumbents per dataset:
+
+        Datastet            Metric      Explanation                     (Index of Arch, Accuracy)
```
Datastet -> Dataset

KEggensperger commented on hpobench/benchmarks/nas/nasbench_1shot1.py:
```diff
@@ -0,0 +1,309 @@
+
+We recommend using the containerized version of this benchmark.
+If you want to use this benchmark locally (without running it via the corresponding container),
+you need to perform the following steps.
```
Please add a note, that this is a very memory intensive benchmark

PhMueller replied:
Since it does rely on the nasbench101, and not on the 201 as I first thought (sorry :-) ), it is not that memory-consuming any more.

KEggensperger replied:
Okay, sounds even better

PhMueller committed:
remove unnecessary requirements + fix typo

KEggensperger commented on extra_requirements/nasbench_1shot1.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "nasbench_1shot1": ["tensorflow==1.15.0","matplotlib","seaborn", "networkx", "tqdm"]
```
Sorry, I realize this only now: But why do we need matplotlib and seaborn? We don't do any plotting in the container, do we?

PhMueller replied:
Jep. That's right. But Arber's code needs it and there is no setup.py in his code. So, we need to install the necessary requirements here.

KEggensperger reviewed (approved):
See my question. Other than that looks good and can be merged.
